File: algorithms/Fraction Addition and Subtraction/Fraction Addition and Subtraction.js
var fractionAddition = function(expression) {
	const fractions = expression.split(/[+-]/).filter(Boolean);
	const operator = expression.split(/[0-9/]/).filter(Boolean);
	expression[0] !== '-' && operator.unshift('+');

	const gcd = (a, b) => b === 0 ? a : gcd(b, a % b);
	const lcm = fractions.reduce((result, fraction) => {
		const denominator = fraction.split('/')[1];
		return result * denominator / gcd(result, denominator);
	}, 1);

	const molecularSum = fractions.reduce((total, fraction, index) => {
		const [molecular, denominator] = fraction.split('/');
		const multiple = lcm / denominator * (operator[index] === '+' ? 1 : -1);
		return total + molecular * multiple;
	}, 0);

	const resultGcd = gcd(Math.abs(molecularSum), lcm);
	return `${molecularSum / resultGcd}/${lcm / resultGcd}`;
};

File: algorithms/Fraction Addition and Subtraction/Fraction Addition and Subtraction.java
class Solution {
    //Great Common Divisor 
    private int gcd(int a,int b) {
        int r=b%a;
        if (r==0) return a;
        return gcd(r,a);
    }
    private int[] add(int[] a,int[] b){
        if(a[0]==0) return b;
        if(b[0]==0) return a;
        int[] ret=new int[]{a[0]*b[1]+b[0]*a[1],a[1]*b[1]};
        if(ret[0]==0){
           ret[1]=1;
           return ret; 
        }
        int sign=ret[0]<0?-1:1;
        ret[0]*=sign;
        int g=gcd(Math.min(ret[0],ret[1]),Math.max(ret[0],ret[1]));
        if(g>1) {
           ret[0]/=g;
           ret[1]/=g; 
        }
        ret[0]*=sign;
        return ret;
    }
    public String fractionAddition(String expression) {
        return Stream.of(
        expression.replaceAll("\\-","+-").split("\\+")).filter(s->s.length()>0).map(
        s->s.split("/")).map(
        a->new int[]{Integer.parseInt(a[0]),Integer.parseInt(a[1])}).reduce((a,b)->add(a,b)).map(p->p[0]+"/"+p[1]).get();
       
    }
}
